src/main/core/manager.rs

```diff
@@ -317,26 +374,68 @@ impl<'a> Manager<'a> {
                         state.current = display_time;
                     });
 
-                // log a heartbeat message every 'heartbeat_interval' amount of simulated time
-                if let Some(heartbeat_interval) = heartbeat_interval {
-                    if window_start > last_heartbeat + heartbeat_interval {
-                        last_heartbeat = window_start;
-                        self.log_heartbeat(window_start);
+                // run the events
+                scheduler.scope(|s| {
+                    s.run_with_data(
+                        &thread_next_event_times,
+                        move |_, hosts, next_event_time| {
+                            let mut next_event_time = next_event_time.borrow_mut();
+
+                            worker::Worker::reset_next_event_time();
+                            worker::Worker::set_round_end_time(window_end);
+
+                            while let Some(host) = hosts.next() {
```

Comment:
Each thread is given its own iterator. This is supposed to be a little abstract, in the sense that it's up to the scheduler to figure out where the next host comes from. For example in the thread-per-host scheduler, each thread contains a thread-local with its host. In the thread-per-core scheduler, the thread has a host queue, but it can steal from other threads' host queues as well.